userspace/libsinsp/container.cpp

```diff
@@ -151,9 +151,45 @@ bool sinsp_container_manager::resolve_container(sinsp_threadinfo* tinfo, bool qu
 	// Also possibly set the category for the threadinfo
 	identify_category(tinfo);
 
+	// Double check tinfo uid and gid and eventually
+	// add them to usergroup_manager
+	check_user_group(tinfo);
+
 	return matches;
 }
 
+// Set uid checking whether it is an unknown (thus new) user.
+// If it is a new user on a container, notify it.
+// refresh_host_users_groups_list() automatically notifies any new/removed user/group
+void sinsp_container_manager::check_user_group(sinsp_threadinfo* tinfo) {
+	scap_userinfo *user = m_inspector->m_usergroup_manager.get_user(tinfo->m_container_id, tinfo->m_uid);
+	if (!user)
+	{
+		if (tinfo->m_container_id.empty())
+		{
+			// HOST: refresh list from scap; this will eventually generate notifications for any new user/group
+			m_inspector->m_usergroup_manager.refresh_host_users_groups_list();
```

Comment:
You're right of course.
As i said above, this could be fixed by just calling getpwuid and getgrgid for new users/groups just once, for host.
And in case we fail to retrieve info, storing it with "<NA>" like we are doing for containers.

The issue here is how to eventually discard host users; at the moment, we have `sync_host_users_groups` called by sinsp once per minute, that internally calls `scap_refresh_userlist`.  
This is needed to avoid that eg: uid 1000 stores "federico" information, even if it gets deleted and a new user is created.
That would of course be a bug because we are printing wrong informations (no info is better then none).

Instead of calling `scap_refresh_userlist`, we might just clear the host uid/gid maps, and add users as they come from new threadinfos.
WDYT?

Reply:
I think we need to have a map in scap, and eventually refresh it with `getpwuid` and `getgrgid`.

Reply:
Why a map? @deepskyblue86 
Scap only knows about host users/groups.

Btw:
> Instead of calling scap_refresh_userlist, we might just clear the host uid/gid maps, and add users as they come from new threadinfos.

Issue would be that we would lose existing threadinfos user/group informations using this approach.

Reply:
I meant uid / gid based map instead of an array to avoid `scap_refresh_userlist`.

Reply:
Yeah fact is, scap does not really need to be updated (scap list is only used to keep backward compatibility with current scap files that expect the list of users/groups to be stored as an initial header in the scap file).

I think we could just load initial list for host in scap for backward compatibility, and then do the "dynamic" part in sinsp; that is what i'm doing right now.
Only issue remaining is understanding how to catch when an user is removed on host; on containers it does not make any difference at all, because we only store `uid` and `gid` in any case.

Reply:
The only way I see is to periodically purge the whole user/group map. This falls under cache invalidation so will surely be trivial ;)

Reply:
@gnosek  you mean this approach?
> Instead of calling scap_refresh_userlist, we might just clear the host uid/gid maps, and add users as they come from new threadinfos.

My only fear is losing running-threads uid/gid info; i mean the uid and gid will still be there, but we won't be able to access any information other than those because we cleared the mapping.

A better thing we could do is to loop all threadinfos and remove any user/group whose uid/gid is not associated to any threadinfo.

Reply:
That's just bad in a different way (which means it might be better for us). Consider:

* uid 1000 is called "someuser"
* we see a process under uid=1000, hence "someuser"
* the user is deleted and replaced with "anotheruser" (still with uid 1000)
* another process runs as uid=1000

What do we expect to see as the user name for both processes?

To return the expected(? at least for me) values (i.e. "someuser" for proc 1 and "anotheruser" for proc 2), we'd probably have to store refcounted strings along with uids/gids.

Reply:
You are right; i think entirely purging the maps every couple of minutes could  be a good best-effort workaround.

Reply:
Another solution for your proposed case, is to store inside threadinfo the full {user,group}info struct. 
This way, each threadinfo has its own full user,group informations.
I think it is not good to store additional strings (ie: user name, user home, user shell, group name) for each threadinfo though...

Reply:
Latest commit implements this behavior; mind to take a look? @deepskyblue86 @gnosek 
Thank you both!